### PR TITLE
[8.0] New module: Purchase Order Generator

### PR DESCRIPTION
**Pull request by Ehtaga**
This module allows the user to configure the reception of products over a
certain period of time.

It features the configuration part and also a wizard that creates the purchase
order lines based on the configurator.

_Some screenshots:_

Create a configuration model, for instance a laying cycle:

po_generator_config

You can access the wizard either by the menu, either when receiving products.
For example, when receiving 12,000 laying hens:

po_generator_transfer
po_generatorwizard

The purchase order automatically generated:

po_generator_result

### CODE REVIEW

File: purchase_order_generator/__init__.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+#
+#    You should have received a copy of the GNU Affero General Public License
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+##############################################################################
+
+from . import (
+    models,
+    wizard,
+)
```
Comment (nhomar) [inline purchase_order_generator/__init__.py:26]: please use a from . import x per line (it helps us to migrate stuff in the future if all make this kind of things equally)

File: purchase_order_generator/wizard/__init__.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+#
+#    You should have received a copy of the GNU Affero General Public License
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+##############################################################################
+
+from . import (
+    purchase_order_generator,
+    stock_transfer_details,
+)
```
Comment (nhomar) [inline purchase_order_generator/wizard/__init__.py:26]: same than one.

File: purchase_order_generator_analytic/README.rst [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+- Choose the model you just created
+- Define a starting date, a partner, a destination location, an analytic
+account and an initial quantity
+- Validate to generate the purchase order lines
+
+For further information, please visit:
+
+* https://www.odoo.com/forum/help-1
+
+Known issues / Roadmap
```
Comment (nhomar) [inline purchase_order_generator_analytic/README.rst:42]: can you add the images and the functional description here or in the doc/index.rst? (you decide) I think it is goos and helps for future usage.

File: purchase_order_generator_revision/static/description/icon.svg [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
...MXFRWfpVNAkEdBiAFru9V4eBc/5K5VKuHz5cls0USwWTz311FP/sn///kMAfo2W1Ped yndjtTAA0PINirOzs+cOHz48C+DXO3bs2EETSIVCAbquI5VKQZZlXzPAfjHciwF46p/uU8xOJ+u6 rp48efK1u++++9/OnDnzUwAziHjKtpvoRyfQD+x4JwHcdvDgwb+nCSRCWm8j3XLLLUsSSOxv1gn0 2kuQgtbL5/NOaEfbm5+f/8Wjjz763NTU1K/QSugYaBG97wlPsZo0AAsCoAjg0tGjR3O5XO5nu3fv 3pZIJDbYto1isQhVVT21AWsC/DaGAFom5vLly22hnaZp8z/84Q/3f+tb33plbm7uZwDyWEVSz2K1 aQA36Pg3A9j14osv3vXpT3/6HrrmIBaLYWJiAjfffHPbRWx2jre3MA3tqNQ7nRGC8+fPv7V///5D U1NT/wtgFi3CryqpZ7FaNQALglZy5ZN33nmncOXKlV/u3LlzQyqV2mzbrR1Dy+VyWwKJ1QC8xSiV SgXT09NQ1WvRm6qq51955ZV9jzzyyMGrUl/CKpV6FqtdA7Cg97IJwK4nnnjiL++4444HZFl2VnxM TExg06ZNKJfLAFqS7n6BNZfLoVKpOGWmaaqnTp06dP/9978J4JcAKlgF3r0o1hIDANfuRwHwhd27 d29+/PHH79+4ceNf0wqJRALDw8POHACdZCoUCsjn820vXxSLxVPPP//8c8eOHfsVgI/QiunXDPGB tWECeLAAZHO5XPmNN944NzIy8tH27dt3ybKcort/0IWplmUhm822zdWbpqn+5Cc/efree+/9/tX8 fc9W6PQaa00DsKD3lgDwF7t37775oYce+ht2zQEvVzAzM/PTffv2vTw1NfUbABdwzckD1hjxgbWr ASgIWgSczuVyhSNHjuRs2/7Vzp07/yQej48A15hA07T5Y8eO7f/2t7/9H3Nzc+8jwpcv+hlrWQOw oPcZB/AFAJnXXnvt77Zv3/41WZZTZ86cOfT1r3/9EFp2/hOsQVvvheuFAYBr9yoB2ADgs1/96lcn AODtt98+C+AkWgsyKfHXNOEpVv2eQSFACWqhlcBZIIQMNhoNCa08goXrQOW7cT1pABasNgDaJf66 IT5w/TIABXv/1xXhb+AGAAD/D0UpzXih8+WrAAAAAElFTkSuQmCC "
```
Comment (nhomar) [inline purchase_order_generator_revision/static/description/icon.svg:59]: You do not need to add images on bas64 anymore you can add them as links to your static folder in the module.

File: purchase_order_generator_revision/static/description/index.html [hunk 5]
```diff
@@ -0,0 +1,124 @@
+<section class="oe_container">
+    <div class="oe_row oe_spaced">
+        <div class="oe_span12">
+            <h2 class="oe_slogan">Module name</h2>
+            <p>This module was written to extend the functionality of ... to support ... and allow you to ...</p>
+        </div>
```
Comment (nhomar) [inline purchase_order_generator_revision/static/description/index.html:6]: Update this file with your actual information, if not in odoo.com/apps will be shown incorrectly.

File: purchase_order_generator_revision/tests/__init__.py [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#    GNU Affero General Public License for more details.
+#
+#    You should have received a copy of the GNU Affero General Public License
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+##############################################################################
+
+from . import (
+    test_purchase_order_generator_revision,
```
Comment (nhomar) [inline purchase_order_generator_revision/tests/__init__.py:24]: 1 line

**Comment by nhomar** [comment]
Falling appart of the comments of form I have a functional question:

Conceptually this module helps you to plan line by line in a purchase order when it will be received? sending a purchase order for every set of element in your reception plans no?

Can you extend a little teh funtional scope, because I understand what it is doing, but what is not clear for me very well is "Why" do I need this..... It looks very promising module.

File: purchase_order_generator_revision/wizards/purchase_order_generator_revision.py [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+    @api.depends("order_id")
+    def _compute_target_received_qty(self):
+        """
+        Compute quantity stock moves from received (done) pickings
+        Compute target from available stock moves.
+        """
+        self.received_qty = (
+            sum(
+                sum(l.product_uom_qty for l in p.move_lines)
+                for p in self.order_id.picking_ids
```
Comment (duanyp1991) [inline purchase_order_generator_revision/wizards/purchase_order_generator_revision.py:79]: Check first if `self.order_id` is  False or not.

Comment (pedrobaeza) [inline purchase_order_generator_revision/wizards/purchase_order_generator_revision.py:79]: There's no need for that with new API: an empty list will be received if sale.order_id is not set.

Comment (duanyp1991) [inline purchase_order_generator_revision/wizards/purchase_order_generator_revision.py:79]: Just tested and yes, true, thanks. :)

**Comment by pedrobaeza** [comment]
No answer on reviewers' comments, so I close. Feel free to reopen if there's still interest in this and comments are attended/replied.

**Pull request closed without merging**